### ISSUE TITLE
PacketDecodeError with SSH_FX_OK

### ISSUE THREAD

**Issue by wmiltenburg** [ronf/asyncssh#224]
Hi,

First of all, I'm not sure if the device I'm connecting with has an incorrect SFTP implementation, so sorry in advance if this is not a bug/issue in asyncssh.

When I'm copying a file with asyncssh using the SFTP client, I get a PacketDecodeError when the connection is closed. After careful debugging, I noticed that the server of a specific device returns a status response SFTP message, with status status code 0/SSH_FX_OK, but without an error message.

If you add a print statement in `_process_status` (sftp.py) to print the payload, you would see this:
```b'e\x00\x00\x00\x03\x00\x00\x00\x00' (request id, error / status code)```

If I look at the draft RFC (https://tools.ietf.org/html/draft-ietf-secsh-filexfer-13) it is not really conclusive if one may expect an error message with SSH_FX_OK.

As `_process_status` tries to fetch the error message, while it is not included in the message, it will result in the PacketDecodeError exception.

A quick workaround for this problem is to add a check in _process_status if we've received an SSH_FX_OK message and just return None.

The workaround for me works, but I'm not sure if this is the correct behaviour or just a workaround for a server side related bug.

Tested assyncssh version: 1.15 - 1.17.0.

With kind regards,
Wouter

**Comment by ronf** [comment]
Based on the RFC you have linked here, I'd say it is pretty clear that the SSH_FXP_STATUS error message and language tag are required argument for all FXP_STATUS responses, regardless of whether the error code is FX_OK or not. Section 9.1 of this doc describes the message as:

```
   The format of the data portion of the SSH_FXP_STATUS response is as
   follows:

       byte   SSH_FXP_STATUS
       uint32 request-id
       uint32 error/status code
       string error message (ISO-10646 UTF-8 [RFC-2279])
       string language tag (as defined in [RFC-1766])
         error-specific data

   request-id
      The 'request-id' specified by the client in the request the server
      is responding to.

   error/status code
      Machine readable status code indicating the result of the request.
      Error code values are defined below.  The value SSH_FX_OK
      indicates success, and all other values indicate failure.

      Implementations MUST be prepared to receive unexpected error codes
      and handle them sensibly (such as by treating them as equivalent
      to SSH_FX_FAILURE).  Future protocol revisions may add additional
      error codes without changing the version number.

   error message
      Human readable description of the error.

   language tag
      'language tag' specifies the language the error is in.

   error-specific data
      The error-specific data may be empty, or may contain additional
      information about the error.  For error codes that send error-
      specific data, the format of the data is defined below.
```

These fields might usually be blank in the FX_OK case, but they should always be included. When blank, each of these would turn into a 32-bit value of \x00\x00\x00\x00 to encode a length of 0 after the FX_OK status code, making the minimal message a total of 17 bytes long.

Only the "error-specific data" would vary from one response code to another. Everything before that should be in every FXP_STATUS response, including FX_OK responses, and this has been the case for all of the SFTP servers I have tested against. If you don't mind my asking, which server is it that is sending an FX_OK without an error message or language tag?

I could potentially make AsyncSSH tolerate this error, automatically filling in empty string values for the message and tag, and in the case of FX_OK that shouldn't break anything, but I'd like to find out a bit more information about the server so that I can document this appropriately.

**Comment by wmiltenburg** [comment]
The part in the RFC that I struggled with was:
```
error/status code
      Machine readable status code indicating the result of the request.
      Error code values are defined below.  The value SSH_FX_OK
      indicates success, and all other values indicate failure.
```

It's mentioned that SSH_FX_OK is not a failure and 'error message' could be interpreted as: an error message that should only be added in case of a failure. The 'language tag' description also mentions that we're in an error/failure situation.

About the specific server implementation: I can't access the server now (on-going VPN migration which locked me out). Sorry. I'll share this later on.

**Comment by ronf** [comment]
> It's mentioned that SSH_FX_OK is not a failure and 'error message' could be interpreted as: an error message that should only be added in case of a failure. The 'language tag' description also mentions that we're in an error/failure situation.

I guess I can see how someone might read it that way, but none of the implementations I've run across in my interoperability testing have had this problem. All of them always include the two additional strings (error message and language tag) for both FX_OK and the error cases. Usually, these strings are empty on FX_OK, but they are still present.

Here's a potential fix for this:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 0bbf7b1..3de5e9a 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -1067,11 +1067,20 @@ class SFTPClientHandler(SFTPHandler):
 
         code = packet.get_uint32()
 
-        try:
-            reason = packet.get_string().decode('utf-8')
-            lang = packet.get_string().decode('ascii')
-        except UnicodeDecodeError:
-            raise SFTPError(FX_BAD_MESSAGE, 'Invalid status message') from None
+        if packet:
+            try:
+                reason = packet.get_string().decode('utf-8')
+                lang = packet.get_string().decode('ascii')
+            except UnicodeDecodeError:
+                raise SFTPError(FX_BAD_MESSAGE,
+                                'Invalid status message') from None
+        else:
+            # Some servers may not always send reason and lang (usually
+            # when responding with FX_OK). Tolerate this, automatically
+            # filling in empty strings for them if they're not present.
+
+            reason = ''
+            lang = ''
 
         packet.check_end()
 
diff --git a/tests/test_sftp.py b/tests/test_sftp.py
index b47f64b..3812154 100644
--- a/tests/test_sftp.py
+++ b/tests/test_sftp.py
@@ -2196,6 +2196,22 @@ class _TestSFTP(_CheckSFTP):
             with self.assertRaises(SFTPError):
                 yield from sftp.open('file')
 
+    @sftp_test
+    def test_short_ok_response(self, sftp):
+        """Test sending an FX_OK response without a reason and lang"""
+
+        @asyncio.coroutine
+        def _short_ok_response(self, pkttype, pktid, packet):
+            """Send an FX_OK response missing reason and lang"""
+
+            # pylint: disable=unused-argument
+
+            self.send_packet(FXP_STATUS, pktid, UInt32(pktid), UInt32(FX_OK))
+
+        with patch('asyncssh.sftp.SFTPServerHandler._process_packet',
+                   _short_ok_response):
+            self.assertIsNone((yield from sftp.mkdir('dir')))
+
     @sftp_test
     def test_malformed_realpath_response(self, sftp):
         """Test receiving malformed realpath response"""
```

It allows the error message and language tag to be left out of any FXP_STATUS response, regardless of the error/status code, as the change is a bit simpler by doing that. It'll fill in empty strings for these fields when they are left out, and that's what will be provided in the SFTPError object that is raised in cases where the status is not FX_OK. I could imagine that possibly happening with an error code of FX_EOF -- typically, that includes an empty error message.

I haven't checked this in yet, but I have it ready to go, if it looks reasonable to you.

**Comment by wmiltenburg** [comment]
This is a nice fix. I can try it tomorrow I you want, but looking at the code it should solve the problem I'm facing.

Thanks for all the help and the great project!

**Comment by ronf** [comment]
Happy to help - I'm glad AsyncSSH is working for you.

If you run into any issues with it, please let me know. Thanks for the report!

**Comment by wmiltenburg** [comment]
Just tested it and it works. Thanks!

**Comment by ronf** [comment]
This change is now available in AsyncSSH 1.17.1.
